### ISSUE TITLE
[BUG] DTE and build failures 

### ISSUE THREAD

**Issue by benpsnyder** [nx-dotnet/nx-dotnet#669]
### Current Behavior

I have a repo with about 300 projects. Let's say 10 are @nx-dotnet app services, and 30 are supporting libs.  When building on a single VM, builds complete fine. When running on CI without DTE, builds complete fine.

However when using DTE, builds do not complete fine. An example of the output:
![image](https://user-images.githubusercontent.com/707567/229299443-64a4d398-b8ae-452e-96ba-36ea6fb65ee9.png)

Perhaps we could improve docs on support for DTE in a large repo with lots of @nx-dotnet packages?

Would it help to set this to `false`?
```zsh
"options": {
  ...
  "noDependencies": false
  ...
},
```

Thank you @AgentEnder in advance for your input!

### Expected Behavior

Builds are expected to complete with DTE the same as they complete on a single VM.

### Steps to Reproduce

Execute codebase with many @nx-dotnet projects with DTE

### Nx Report

```shell
nx report
>  NX  Falling back to ts-node for local typescript execution. This may be a little slower.
  - To fix this, ensure @swc-node/register and @swc/core have been installed

 >  NX   Report complete - copy this into the issue template

   Node : 19.8.1
   OS   : darwin arm64
   pnpm : 8.1.0

   nx (global)             : 15.9.2
   nx                      : 15.9.0
   @nrwl/js                : 15.9.0
   @nrwl/jest              : 15.9.0
   @nrwl/linter            : 15.9.0
   @nrwl/workspace         : 15.9.0
   @nrwl/angular           : 15.9.0
   @nrwl/cli               : 15.9.0
   @nrwl/cypress           : 15.9.0
   @nrwl/devkit            : 15.9.0
   @nrwl/eslint-plugin-nx  : 15.9.0
   @nrwl/nest              : 15.9.0
   @nrwl/node              : 15.9.0
   @nrwl/storybook         : 15.9.0
   @nrwl/webpack           : 15.9.0
   @nrwl/nx-cloud          : 15.3.3
   typescript              : 5.0.2
   ---------------------------------------
   Community plugins:
   @nestjs/azure-storage                : 3.0.0
   @alyle/ui                            : 13.4.0
   @fortawesome/angular-fontawesome     : 0.12.1
   @jscutlery/semver                    : 2.30.1
   @ngrx/schematics                     : 15.4.0
   @nrwl/rspack                         : 0.158.6
   @nstudio/xplat                       : 15.0.3
   @nx-dotnet/core                      : 1.21.0
   @nx-dotnet/nxdoc                     : 1.21.0
   @nx-enterprise/nxe-devcontainer      : 0.0.3
   @nx-plus/docusaurus                  : 15.0.0-rc.0
   @nx-tools/nx-container               : 4.0.3
   @progress/kendo-angular-buttons      : 8.2.2
   @progress/kendo-angular-charts       : 7.3.2
   @progress/kendo-angular-common       : 3.2.2
   @progress/kendo-angular-dateinputs   : 7.1.6
   @progress/kendo-angular-dialog       : 7.1.5
   @progress/kendo-angular-dropdowns    : 7.2.4
   @progress/kendo-angular-excel-export : 5.0.2
   @progress/kendo-angular-gauges       : 5.0.1
   @progress/kendo-angular-grid         : 7.4.2
   @progress/kendo-angular-inputs       : 10.1.2
   @progress/kendo-angular-intl         : 4.1.1
   @progress/kendo-angular-l10n         : 4.0.1
   @progress/kendo-angular-label        : 4.0.2
   @progress/kendo-angular-layout       : 7.2.0
   @progress/kendo-angular-notification : 4.0.1
   @progress/kendo-angular-pager        : 4.0.6
   @progress/kendo-angular-pdf-export   : 4.0.1
   @progress/kendo-angular-popup        : 5.0.2
   @progress/kendo-angular-progressbar  : 3.1.2
   @progress/kendo-angular-scrollview   : 5.0.2
   @progress/kendo-angular-tooltip      : 4.1.1
   @progress/kendo-angular-treeview     : 7.1.5
   @progress/kendo-angular-upload       : 9.0.4
   @progress/kendo-popup-common         : 1.9.2
   @srleecode/domain                    : 15.2.5
   @storybook/angular                   : 6.5.16
   @syncfusion/ej2-angular-base         : 21.1.37
   @syncfusion/ej2-angular-buttons      : 21.1.37
   @syncfusion/ej2-angular-calendars    : 21.1.37
   @syncfusion/ej2-angular-dropdowns    : 21.1.37
   @syncfusion/ej2-angular-grids        : 21.1.37
   @syncfusion/ej2-angular-inputs       : 21.1.37
   @syncfusion/ej2-angular-lists        : 21.1.37
   @syncfusion/ej2-angular-navigations  : 21.1.37
   @syncfusion/ej2-angular-pivotview    : 21.1.37
   @syncfusion/ej2-angular-popups       : 21.1.37
   @syncfusion/ej2-angular-splitbuttons : 21.1.37
   eslint-plugin-ngrx                   : 2.1.4
   ngx-spinner                          : 15.0.1
```

### nx.json

```json
{
  "$schema": "./node_modules/nx/schemas/nx-schema.json",
  "affected": {
    "defaultBase": "develop"
  },
  "cli": {
    "analytics": "72c4c923-4113-4bf2-923a-10a88abdd92f"
  },
  "generators": {
    "@nrwl/angular:application": {
      "e2eTestRunner": "cypress",
      "linter": "eslint",
      "style": "scss",
      "unitTestRunner": "jest"
    },
    "@nrwl/angular:component": {
      "style": "scss"
    },
    "@nrwl/angular:library": {
      "linter": "eslint",
      "strict": true,
      "unitTestRunner": "jest"
    },
    "@nx-dotnet/core:app": {
      "language": "C#",
      "tags": "type:api",
      "template": "webapi",
      "testTemplate": "xunit"
    },
    "@nx-dotnet/core:lib": {
      "language": "C#",
      "template": "classlib",
      "testTemplate": "xunit"
    }
  },
  "namedInputs": {
    "default": ["{projectRoot}/**/*", "sharedGlobals"],
    "production": ["default"],
    "sharedGlobals": ["{workspaceRoot}/workspace.json", "{workspaceRoot}/nx.json", "{workspaceRoot}/tsconfig.base.json"]
  },
  "plugins": ["@nx-dotnet/core"],
  "targetDefaults": {
    "build": {
      "dependsOn": ["^build", "codegen", "^codegen"],
      "inputs": ["production", "^production"]
    },
    "build-storybook": {
      "inputs": ["default", "^production", "{workspaceRoot}/.storybook/**/*"]
    },
    "e2e": {
      "inputs": ["default", "^production"]
    },
    "lint": {
      "inputs": ["default", "{workspaceRoot}/.eslintrc.json"]
    },
    "test": {
      "inputs": ["default", "^production", "{workspaceRoot}/jest.preset.js"]
    }
  },
  "tasksRunnerOptions": {
    "default": {
      "options": {
        "accessToken": "<token>",
        "cacheableOperations": ["build", "test", "lint", "e2e", "codegen", "swagger"],
        "canTrackAnalytics": true,
        "parallel": 8,
        "showUsageWarnings": true
      },
      "runner": "@nrwl/nx-cloud"
    }
  }
}
```

**Comment by AgentEnder** [comment]
It should work with DTE - I'll spin up a test repo next week to look into this more thoroughly 

In the meantime, can you grab the build target configuration and the build.directory.props contents for me?

**Comment by benpsnyder** [comment]
Sure thing.  My builds were passing 30% of the time when `--parallel=3` and when I increased to `--parallel=8` I started getting a higher build success rate.  I have all nx-dotnet projects set to `noDependencies: false` in order to get it to work.

I went through PRs for recent releases and came across this: https://github.com/nx-dotnet/nx-dotnet/pull/596 

I'm working to add intermediates in all the projects definitions now.
![image](https://user-images.githubusercontent.com/707567/229385359-ab4f0156-70f2-43f5-8019-68093450d8cc.png)

My updated config:
```c#
<Project>
  <PropertyGroup>
    <!-- Output path configuration -->
    <RepoRoot>$([System.IO.Path]::GetFullPath('$(MSBuildThisFileDirectory)'))</RepoRoot>
    <ProjectRelativePath>$([MSBuild]::MakeRelative($(RepoRoot), $(MSBuildProjectDirectory)))</ProjectRelativePath>
    <BaseOutputPath>$(RepoRoot)dist/$(ProjectRelativePath)</BaseOutputPath>
    <OutputPath>$(BaseOutputPath)</OutputPath>
    <BaseIntermediateOutputPath>$(RepoRoot)dist/intermediates/$(ProjectRelativePath)/obj</BaseIntermediateOutputPath>
    <IntermediateOutputPath>$(BaseIntermediateOutputPath)</IntermediateOutputPath>
    <AppendTargetFrameworkToOutputPath>true</AppendTargetFrameworkToOutputPath>
  </PropertyGroup>
  <PropertyGroup>
    <RestorePackagesWithLockFile>false</RestorePackagesWithLockFile>
    <LangVersion>latest</LangVersion>
    <TargetFramework>net6.0</TargetFramework>
    <Nullable>enable</Nullable>
    <GenerateDocumentationFile>true</GenerateDocumentationFile>
    <TreatWarningsAsErrors>false</TreatWarningsAsErrors>
    <DocumentationFile>$(OutputPath)$(AssemblyName).xml</DocumentationFile>
    <ImplicitUsings>enable</ImplicitUsings>
  </PropertyGroup>
</Project>
```

Newly generated `project.json` from the generated contain
```json
{
...
"outputs": ["{workspaceRoot}/dist/apps/testest", "{workspaceRoot}/dist/intermediates/apps/testest"]
...
}
```
But the majority of my projects in the repo didn't get this from migrations.

I'll try to do more legwork for you here to save you time (I'll post back), and make a sample repo if I can to reproduce. Workload is heavy right now though so I can't promise that.

**Comment by benpsnyder** [comment]
Log message: ` Unable to use package assets cache due to I/O error. This can occur when the same project is built more than once in parallel. Performance may be degraded, but the build result will not be impacted.`

Sample Nx Cloud output with some redactions
```sh
/home/<USER>/.azure-devops-agents/az-lxweb05-agent-3/_work/1/s/libs/core/part<USER>/common/Services/AESService.cs(263,42): warning CS1591: Missing XML comment for publicly visible type or member 'EdgePart<USER>BackendUtilityAesService.GetAESSubmissionLogs(int?, DateTime?, DateTime?, string, int, int)' [/home/<USER>/.azure-devops-agents/az-lxweb05-agent-3/_work/1/s/libs/core/part<USER>/common/Acme.Core.Part<USER>.Common.csproj]
  Acme.Core.Part<USER>.Common -> /home/<USER>/.azure-devops-agents/az-lxweb05-agent-3/_work/1/s/dist/libs/core/part<USER>/common/net6.0/Acme.Core.Part<USER>.Common.dll
  Acme.Core.Reporting.Common -> /home/<USER>/.azure-devops-agents/az-lxweb05-agent-3/_work/1/s/dist/libs/core/reporting/common/net6.0/Acme.Core.Reporting.Common.dll
  Acme.Core.Shared.DbConnections.Hyperscale -> /home/<USER>/.azure-devops-agents/az-lxweb05-agent-3/_work/1/s/dist/libs/core/shared/db-connections/hyperscale/net6.0/Acme.Core.Shared.DbConnections.Hyperscale.dll
  Acme.Core.Connectivity.Common -> /home/<USER>/.azure-devops-agents/az-lxweb05-agent-3/_work/1/s/dist/libs/core/connectivity/common/net6.0/Acme.Core.Connectivity.Common.dll
  Acme.Core.Connectivity.Chi -> /home/<USER>/.azure-devops-agents/az-lxweb05-agent-3/_work/1/s/dist/libs/core/connectivity/chi/net6.0/Acme.Core.Connectivity.Chi.dll
/home/<USER>/.azure-devops-agents/az-lxweb05-agent-3/_work/_tool/dotnet/sdk/6.0.407/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.targets(181,5): error MSB4018: The "GenerateDepsFile" task failed unexpectedly. [/home/<USER>/.azure-devops-agents/az-lxweb05-agent-3/_work/1/s/libs/core/connectivity/XYZ/Acme.Core.Connectivity.XYZ.csproj]
/home/<USER>/.azure-devops-agents/az-lxweb05-agent-3/_work/_tool/dotnet/sdk/6.0.407/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.targets(181,5): error MSB4018: System.IO.IOException: The process cannot access the file '/home/<USER>/.azure-devops-agents/az-lxweb05-agent-3/_work/1/s/dist/libs/core/connectivity/XYZ/net6.0/Acme.Core.Connectivity.XYZ.deps.json' because it is being used by another process. [/home/<USER>/.azure-devops-agents/az-lxweb05-agent-3/_work/1/s/libs/core/connectivity/XYZ/Acme.Core.Connectivity.XYZ.csproj]
/home/<USER>/.azure-devops-agents/az-lxweb05-agent-3/_work/_tool/dotnet/sdk/6.0.407/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.targets(181,5): error MSB4018:    at Microsoft.Win32.SafeHandles.SafeFileHandle.Init(String path, FileMode mode, FileAccess access, FileShare share, FileOptions options, Int64 preallocationSize) [/home/<USER>/.azure-devops-agents/az-lxweb05-agent-3/_work/1/s/libs/core/connectivity/XYZ/Acme.Core.Connectivity.XYZ.csproj]
/home/<USER>/.azure-devops-agents/az-lxweb05-agent-3/_work/_tool/dotnet/sdk/6.0.407/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.targets(181,5): error MSB4018:    at Microsoft.Win32.SafeHandles.SafeFileHandle.Open(String fullPath, FileMode mode, FileAccess access, FileShare share, FileOptions options, Int64 preallocationSize) [/home/<USER>/.azure-devops-agents/az-lxweb05-agent-3/_work/1/s/libs/core/connectivity/XYZ/Acme.Core.Connectivity.XYZ.csproj]
/home/<USER>/.azure-devops-agents/az-lxweb05-agent-3/_work/_tool/dotnet/sdk/6.0.407/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.targets(181,5): error MSB4018:    at System.IO.Strategies.OSFileStreamStrategy..ctor(String path, FileMode mode, FileAccess access, FileShare share, FileOptions options, Int64 preallocationSize) [/home/<USER>/.azure-devops-agents/az-lxweb05-agent-3/_work/1/s/libs/core/connectivity/XYZ/Acme.Core.Connectivity.XYZ.csproj]
/home/<USER>/.azure-devops-agents/az-lxweb05-agent-3/_work/_tool/dotnet/sdk/6.0.407/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.targets(181,5): error MSB4018:    at System.IO.Strategies.FileStreamHelpers.ChooseStrategy(FileStream fileStream, String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, FileOptions options, Int64 preallocationSize) [/home/<USER>/.azure-devops-agents/az-lxweb05-agent-3/_work/1/s/libs/core/connectivity/XYZ/Acme.Core.Connectivity.XYZ.csproj]
/home/<USER>/.azure-devops-agents/az-lxweb05-agent-3/_work/_tool/dotnet/sdk/6.0.407/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.targets(181,5): error MSB4018:    at System.IO.File.Create(String path) [/home/<USER>/.azure-devops-agents/az-lxweb05-agent-3/_work/1/s/libs/core/connectivity/XYZ/Acme.Core.Connectivity.XYZ.csproj]
/home/<USER>/.azure-devops-agents/az-lxweb05-agent-3/_work/_tool/dotnet/sdk/6.0.407/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.targets(181,5): error MSB4018:    at Microsoft.NET.Build.Tasks.GenerateDepsFile.WriteDepsFile(String depsFilePath) [/home/<USER>/.azure-devops-agents/az-lxweb05-agent-3/_work/1/s/libs/core/connectivity/XYZ/Acme.Core.Connectivity.XYZ.csproj]
/home/<USER>/.azure-devops-agents/az-lxweb05-agent-3/_work/_tool/dotnet/sdk/6.0.407/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.targets(181,5): error MSB4018:    at Microsoft.NET.Build.Tasks.GenerateDepsFile.ExecuteCore() [/home/<USER>/.azure-devops-agents/az-lxweb05-agent-3/_work/1/s/libs/core/connectivity/XYZ/Acme.Core.Connectivity.XYZ.csproj]
/home/<USER>/.azure-devops-agents/az-lxweb05-agent-3/_work/_tool/dotnet/sdk/6.0.407/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.targets(181,5): error MSB4018:    at Microsoft.NET.Build.Tasks.TaskBase.Execute() [/home/<USER>/.azure-devops-agents/az-lxweb05-agent-3/_work/1/s/libs/core/connectivity/XYZ/Acme.Core.Connectivity.XYZ.csproj]
/home/<USER>/.azure-devops-agents/az-lxweb05-agent-3/_work/_tool/dotnet/sdk/6.0.407/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.targets(181,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute() [/home/<USER>/.azure-devops-agents/az-lxweb05-agent-3/_work/1/s/libs/core/connectivity/XYZ/Acme.Core.Connectivity.XYZ.csproj]
/home/<USER>/.azure-devops-agents/az-lxweb05-agent-3/_work/_tool/dotnet/sdk/6.0.407/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.targets(181,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskBuilder.ExecuteInstantiatedTask(ITaskExecutionHost taskExecutionHost, TaskLoggingContext taskLoggingContext, TaskHost taskHost, ItemBucket bucket, TaskExecutionMode howToExecuteTask) [/home/<USER>/.azure-devops-agents/az-lxweb05-agent-3/_work/1/s/libs/core/connectivity/XYZ/Acme.Core.Connectivity.XYZ.csproj]
  Acme.Core.Connectivity.Primo -> /home/<USER>/.azure-devops-agents/az-lxweb05-agent-3/_work/1/s/dist/libs/core/connectivity/primo/net6.0/Acme.Core.Connectivity.Primo.dll
```
I updated `project.json` for 50+ projects in the repo.
![image](https://user-images.githubusercontent.com/707567/229387716-63bcce5b-1de0-4272-a688-a123f48a1db7.png)

In the above test, I have `noDependencies: false`

Then I tested `noDependencies: true`
![image](https://user-images.githubusercontent.com/707567/229387913-c5c8b0a3-f827-448f-bdef-3512ab95fddd.png)

And wouldn't you know it .... success!

@AgentEnder could you shed some insight into what's happening here? I'd be glad to send in a PR updating docs for you or some enhancement if it helps.

**Comment by benpsnyder** [comment]
Another reason for my inconsistent build failures is I was using `<TargetFrameworks>` and nx-framework doesn't handle the subsequent `/ref/` folder where Reference Assemblies are stored in the dist output.
![image](https://user-images.githubusercontent.com/707567/229583312-a117dc14-1f50-4af4-99db-38637e432bb8.png)

**Comment by Tungsten78** [comment]
I just recently battled this and came here to enter an issue report.  :)
We would see this issue intermittently prior to adding DTE (depending on cache state) and every time once DTE was added.

I think the root cause is that the build target outputs does NOT include `{sourceRoot}/obj`.   Project references are resolved from here - not dist.   

The workaround was to add `"{sourceRoot}/obj"` to each library's build target `outputs` array.

**Comment by AgentEnder** [comment]
@Tungsten78 when you run build do intermediates not get added to dist? In latest, I don't recall seeing a project level obj folder anymore

**Comment by Tungsten78** [comment]
No intermediates folder in dist, despite seeing it in project.json.   I wasn't sure what this was.

I'm on    @nx-dotnet/core              : 1.19.1

**Comment by AgentEnder** [comment]
@benpsnyder We tend to recommend using `noDependencies` as it lets Nx handle the builds and deps. There's a write up of it here: https://www.nx-dotnet.com/docs/core/guides/incremental-builds

With `noDependencies: false`, Nx will still build the project you requested, and its dependencies. `dotnet` on the other hand, will also be trying to build the deps. So for a simple workspace with something like:

parent-a depends on child-a, child-b, and child-c
child-a and child-b depend on child-c

running `nx build parent-a` with `noDependencies: true` results in:

- Nx builds child-c
	- `dotnet build child-c --no-dependencies` # build C, doesn't have deps so equivalent to w/o noDeps
- Nx builds child-b and child-a in parallel
	- `dotnet build child-a --no-dependencies` # build child A, consume child C from dist
	- `dotnet build child-b --no-dependencies` # build child B, consume C from dist
- Nx builds parent-a
	- `dotnet build parent-a --no-dependencies` # build parent A, consume all children from dist

With `noDependencies: false`, you get something closer to:
- Nx builds child-c 
 	- `dotnet build child-c` # build C, doesn't have deps so equivalent to w noDeps
- Nx builds child-a and child-b in parallel
	- `dotnet build child-a` # build child A, but first build child C
	- `dotnet build child-b` # build child B, but first build child C
- Nx builds parent-a
	- `dotnet build parent-a` # build parent A, but build all the deps first

In scenario 1, we only let .NET build each child once, there's no extra work, but also if child C was a project that took a long time to cache, it could have been pulled down from the cache.

In scenario 2, .NET builds child C an astonishing 4 times. There are some optimizations under the hood of .NET, but regardless that adds up and is one of the reasons the plugin is such a perf boost when configured properly.

----

@Tungsten78 There's some fragments inside Directory.build.props that's purpose is to send build intermediates to dist/intermediates/{projectRoot}, if your intermediates aren't there and you don't update the project level outputs props thats definitely why you'd get this issue.

**Comment by AgentEnder** [comment]
I'm going to close this out, as I truly believe it is working as intended and the issues been quiet for ~2 weeks now. If anyone is still having issues feel free to comment or open a new issue referencing this one.

**Comment by Tungsten78** [comment]
@AgentEnder  - manually adding `BaseIntermediateOutputPath` and `IntermediateOutputPath` to Directory.Build.props solved my issue.   Why they were absent in the first place remains a mystery.
